Basic text renderer, initial PR

andydotxyz opened the pull request:
Works well enough that Fyne can use it for all its font styles and various faces, so here is the first PR :).

Current test result:
![out](https://user-images.githubusercontent.com/294436/216074837-7c698541-4afe-456c-b6d4-bf96cbc58345.png)

whereswaldon commented on render.go:
```diff
@@ -0,0 +1,87 @@
+	Color              color.Color
+}
+
+func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
+	sh := &shaping.HarfbuzzShaper{}
+	in := shaping.Input{
+		Text:     []rune(str),
+		RunStart: 0,
+		RunEnd:   len(str),
+		Face:     face,
+		Size:     fixed.I(int(r.FontSize * r.PixScale)),
+	}
+	out := sh.Shape(in)
+	return r.DrawShapedRunAt(out, img, 0, out.LineBounds.Ascent.Ceil())
+}
+
+func (r *Renderer) DrawStringAt(str string, img draw.Image, x, y int, face fonts.Face) int {
+	sh := &shaping.HarfbuzzShaper{}
+	in := shaping.Input{
+		Text:     []rune(str),
+		RunStart: 0,
+		RunEnd:   len(str),
+		Face:     face,
+		Size:     fixed.I(int(r.FontSize * r.PixScale)),
+	}
+	return r.DrawShapedRunAt(sh.Shape(in), img, x, y)
+}
```
Should these two methods wrap the string to fit within the provided image (as best as possible, anyway)?

andydotxyz replied:
That is a good question. I'm not sure. As we are not passing any sizing info it may overflow on the X or the Y.
From a "canvas" point of view my assumption had been that basic text paint isn't exposing smarts and you would use the run painting variant if you want to do anything clever like wrap or multi-font etc.
As I guess the API wouldn't change perhaps we land this simplistic version and then if it is expected we can wrap later?

whereswaldon replied:
Sure!

whereswaldon commented on render.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
+	sh := &shaping.HarfbuzzShaper{}
```
There are significant efficiency gains from reusing a shaper instead of re-creating one each time you rasterize. Perhaps this should be a field on Renderer, assuming that Renderers are meant to be reused? Similarly, if you do any line wrapping, reusing the wrapper is much more efficient.

whereswaldon replied:
Well, I tried it, and it seems like the raster process totally dominates, so it's a very minor efficiency gain. If you'd like to experiment, here is what I did:

```diff
commit dab5cca71a7e29d5dabd2efba0fb6ebf7bd24aa0
Author: Chris Waldon <christopher.waldon.dev@gmail.com>
Date:   Wed Feb 1 16:47:39 2023 -0500

    render: reuse shaper
    
    Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

diff --git a/render.go b/render.go
index 8f12dca..d6aae68 100644
--- a/render.go
+++ b/render.go
@@ -14,10 +14,10 @@ import (
 type Renderer struct {
 	FontSize, PixScale float32
 	Color              color.Color
+	shaper             shaping.HarfbuzzShaper
 }
 
 func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
-	sh := &shaping.HarfbuzzShaper{}
 	in := shaping.Input{
 		Text:     []rune(str),
 		RunStart: 0,
@@ -25,12 +25,11 @@ func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
 		Face:     face,
 		Size:     fixed.I(int(r.FontSize * r.PixScale)),
 	}
-	out := sh.Shape(in)
+	out := r.shaper.Shape(in)
 	return r.DrawShapedRunAt(out, img, 0, out.LineBounds.Ascent.Ceil())
 }
 
 func (r *Renderer) DrawStringAt(str string, img draw.Image, x, y int, face fonts.Face) int {
-	sh := &shaping.HarfbuzzShaper{}
 	in := shaping.Input{
 		Text:     []rune(str),
 		RunStart: 0,
@@ -38,7 +37,7 @@ func (r *Renderer) DrawStringAt(str string, img draw.Image, x, y int, face fonts
 		Face:     face,
 		Size:     fixed.I(int(r.FontSize * r.PixScale)),
 	}
-	return r.DrawShapedRunAt(sh.Shape(in), img, x, y)
+	return r.DrawShapedRunAt(r.shaper.Shape(in), img, x, y)
 }
 
 func (r *Renderer) DrawShapedRunAt(run shaping.Output, img draw.Image, startX, startY int) int {

commit 08359a8ae889a3240b92dddaaea95e0e4eff2f8f
Author: Chris Waldon <christopher.waldon.dev@gmail.com>
Date:   Wed Feb 1 16:33:38 2023 -0500

    render: add simple benchmark
    
    Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

diff --git a/render_test.go b/render_test.go
index 2edcc0b..01a005b 100644
--- a/render_test.go
+++ b/render_test.go
@@ -41,3 +41,22 @@ func Test_Render(t *testing.T) {
 	png.Encode(w, img)
 	w.Close()
 }
+
+func Benchmark_Render(b *testing.B) {
+	img := image.NewNRGBA(image.Rect(0, 0, 350, 180))
+	draw.Draw(img, img.Bounds(), &image.Uniform{color.White}, image.Point{}, draw.Src)
+
+	data, _ := os.Open("testdata/NotoSans-Regular.ttf")
+	f, _ := font.ParseTTF(data)
+
+	r := &render.Renderer{
+		FontSize: 24,
+		PixScale: float32(2),
+		Color:    color.Black,
+	}
+	str := "Hello! ± सभमन"
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		r.DrawString(str, img, f)
+	}
+}
```

whereswaldon commented on render.go:
```diff
@@ -0,0 +1,87 @@
+)
+
+type Renderer struct {
+	FontSize, PixScale float32
```
Can you document the expected units for these fields?

andydotxyz replied:
Ah yes I forgot the docs sorry! Fixed now

whereswaldon reviewed (commented):
Nice work! A few questions for you inline.

benoitkugler commented:
Sounds good to me, nice work !

Perhaps we could mention that the `DrawString` and `DrawStringAt`  are convenience wrapper to both shape and render. (In practice, I  think that the intermediate shape result would be cached and reused, so that the wrappers are not so often use, right ?)

benoitkugler commented on render.go:
```diff
@@ -0,0 +1,87 @@
+	Color              color.Color
+}
+
+func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
```
Could you document where the first glyph is drawn on `img` ?

andydotxyz replied:
Thanks for that, fixed

andydotxyz commented:
> Sounds good to me, nice work !
> 
> Perhaps we could mention that the `DrawString` and `DrawStringAt` are convenience wrapper to both shape and render. (In practice, I think that the intermediate shape result would be cached and reused, so that the wrappers are not so often use, right ?)

I have added missing docs. I didn't explicitly state this as someone using the API at that level probably doesn't care? Let me know if you feel what I added should be improved or clearer.

benoitkugler commented:
> > Sounds good to me, nice work !
> > Perhaps we could mention that the `DrawString` and `DrawStringAt` are convenience wrapper to both shape and render. (In practice, I think that the intermediate shape result would be cached and reused, so that the wrappers are not so often use, right ?)
> 
> I have added missing docs. I didn't explicitly state this as someone using the API at that level probably doesn't care? Let me know if you feel what I added should be improved or clearer.

Thank you for the documentation.
All good to me now !

whereswaldon commented on render.go:
```diff
@@ -0,0 +1,128 @@
+	FontSize float32
+	// PixScale is used to indicate the pixel density of your output target.
+	// For example on a hi-DPI (or "retina") display this may be 2.0.
+	// Default value is 1.0, meaning
```
I think maybe you meant to write more here?

andydotxyz replied:
Oops, must have got side-tracked. Fixed

whereswaldon commented:
Just today I learned about [this package](https://pkg.go.dev/golang.org/x/image@v0.3.0/vector). I'm curious how you chose the rasterizer that you did. What made it stand out among the options? `github.com/srwiley/rasterx` appears to be much higher-level than what we actually need here, but maybe I'm missing something.

andydotxyz commented:
> Just today I learned about [this package](https://pkg.go.dev/golang.org/x/image@v0.3.0/vector). I'm curious how you chose the rasterizer that you did. What made it stand out among the options? `github.com/srwiley/rasterx` appears to be much higher-level than what we actually need here, but maybe I'm missing something.

I just used one that I knew and was using before. It underpins the SVG parser we use for Fyne apps.
I wasn't sure about it being too high level as the API seemed like a good match (as you can see in this PR).
Do I need to go and implement with a different rasteriser for this PR?
Do we need to do benchmarking of different solutions to get ideal speeds?
As this is just the initial code I would opt for landing and if other solutions prove faster we could switch out rasterx.

whereswaldon commented:
> > Just today I learned about [this package](https://pkg.go.dev/golang.org/x/image@v0.3.0/vector). I'm curious how you chose the rasterizer that you did. What made it stand out among the options? `github.com/srwiley/rasterx` appears to be much higher-level than what we actually need here, but maybe I'm missing something.
> 
> I just used one that I knew and was using before. It underpins the SVG parser we use for Fyne apps. I wasn't sure about it being too high level as the API seemed like a good match (as you can see in this PR). Do I need to go and implement with a different rasteriser for this PR? Do we need to do benchmarking of different solutions to get ideal speeds? As this is just the initial code I would opt for landing and if other solutions prove faster we could switch out rasterx.

I just wanted to know what factors went in to the choice. I'm not asking you to change anything. I'd be interested to know whether we could swap in the one in `x/image`, but I don't need to block over that.

andydotxyz commented:
> I'd be interested to know whether we could swap in the one in x/image,

Good question, I will check it out and see how it fares in comparison.
Thanks for the merge.
